### PR TITLE
Add automatic trace support for Spring WebClient

### PR DESCRIPTION
**Pull request by yhware**
## :scroll: Description

When using `WebClient` in Spring, automatically add span and breadcrumb for tracing.
There's already support for `RestTemplate` (https://docs.sentry.io/platforms/java/guides/spring-boot/performance/instrumentation/automatic-instrumentation/)

## :bulb: Motivation and Context

Resolves: Support for Spring WebClient for creating spans/crumbs #1619

## :green_heart: How did you test it?
Ported all tests that were available for RestTemplate integration.
Also ran the sample spring boot application with my personal DSN and verified it worked as intended.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [ ] I updated the docs if needed
- [x] No breaking changes

### CODE REVIEW

**Review by bruno-garcia — APPROVED** [review]
Looks great! @maciejwalkowiak might have some thoughts as the author of the RestTemplate bits

**Comment by marandaneto** [comment]
looks good, I'll let @maciejwalkowiak review though since its backend bits.
@yhware thanks for that, would you mind working on the broken CI? this includes running `make format` to format files, `make api` to check binary compatibility validator and adding an entry to the changelog.

**Review by marandaneto — APPROVED** [review]
left a comment to fix CI but changes are good LGTM

File: sentry-samples/sentry-samples-spring-boot/src/main/java/io/sentry/samples/spring/boot/SentryDemoApplication.java [hunk 1]
```diff
@@ -1,11 +1,13 @@
 package io.sentry.samples.spring.boot;
 
+import io.sentry.spring.EnableSentry;
```
Comment (maciejwalkowiak) [inline sentry-samples/sentry-samples-spring-boot/src/main/java/io/sentry/samples/spring/boot/SentryDemoApplication.java:3]: This must have accidentally sneaked in. Run `./gradlew spotlessApply` from the command line, to remove all unused imports, re-order imports, format code etc.

File: sentry-spring-boot-starter/build.gradle.kts [hunk 2]
```diff
@@ -60,6 +61,8 @@ dependencies {
     testImplementation(kotlin(Config.kotlinStdLib))
     testImplementation(Config.TestLibs.kotlinTestJunit)
     testImplementation(Config.TestLibs.mockitoKotlin)
+    testImplementation(Config.TestLibs.mockWebserver)
+    testImplementation(Config.Libs.okhttp)
```
Comment (maciejwalkowiak) [inline sentry-spring-boot-starter/build.gradle.kts:65]: `Config.Libs.okHttp` does not specify exact OkHttp version. As a result it loads MockWebServer 4.9.0, but OkHttp 3.x which causes later on compilation errors. 

In the `dependencyManagement` configuration in this file (search for `the<DependencyManagementExtension>().apply {`) add:

```
mavenBom(Config.Libs.okhttpBom)
```
To make sure correct version is used.

**Review by maciejwalkowiak — CHANGES_REQUESTED** [review]
@yhware thanks a lot for this PR! 🤩 Few small things have to be polished to get it merged. Also, make sure to run `./gradlew apiDump` to update files containing public APIs.

File: sentry-spring-boot-starter/build.gradle.kts (hunk 2 above)
Comment (yhware) [reply sentry-spring-boot-starter/build.gradle.kts:66]: Ah.. I ran into this issue and then it was suddenly gone.
Committed the changes in 713b11a4c95b0741d9f5cc34022c1811048c30c8

File: sentry-samples/sentry-samples-spring-boot/src/main/java/io/sentry/samples/spring/boot/SentryDemoApplication.java (hunk 1 above)
Comment (yhware) [reply sentry-samples/sentry-samples-spring-boot/src/main/java/io/sentry/samples/spring/boot/SentryDemoApplication.java]: Yup...my bad. lint quickly fixed this :)

File: sentry-spring/src/main/java/io/sentry/spring/tracing/SentrySpanClientWebRequestFilter.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
+package io.sentry.spring.tracing;
+
+import com.jakewharton.nopen.annotation.Open;
+import io.sentry.*;
```
[the remark this answers is not in the document]
Comment (yhware) [reply sentry-spring/src/main/java/io/sentry/spring/tracing/SentrySpanClientWebRequestFilter.java]: Suggestion applied with 713b11a4c95b0741d9f5cc34022c1811048c30c8 :)

File: sentry-spring/src/main/java/io/sentry/spring/tracing/SentrySpanClientWebRequestFilter.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+@Open
+public class SentrySpanClientWebRequestFilter implements ExchangeFilterFunction {
+  private final @NotNull IHub hub;
+
+  public SentrySpanClientWebRequestFilter(final @NotNull IHub hub) {
+    this.hub = Objects.requireNonNull(hub, "hub is required");
+  }
+
+  @Override
+  public Mono<ClientResponse> filter(@NotNull ClientRequest request, @NotNull ExchangeFunction next) {
```
[the remark this answers is not in the document]
Comment (yhware) [reply sentry-spring/src/main/java/io/sentry/spring/tracing/SentrySpanClientWebRequestFilter.java]: Suggestion applied with https://github.com/getsentry/sentry-java/commit/9126af8b08f58eb67e8f2e5dc3b3ea7f66afc28b#diff-b2ff2c453fd18be47feeae886608e7b4a561f2bcd6189ec23239b77dd322b90fR27 :)

File: sentry-spring-boot-starter/src/test/kotlin/io/sentry/spring/boot/SentrySpanWebClientCustomizerTest.kt [hunk 5]
```diff
@@ -0,0 +1,227 @@
+package io.sentry.spring.boot
+
+import com.nhaarman.mockitokotlin2.check
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.verify
+import com.nhaarman.mockitokotlin2.whenever
+import io.sentry.*
```
[the remark this answers is not in the document]
Comment (yhware) [reply sentry-spring-boot-starter/src/test/kotlin/io/sentry/spring/boot/SentrySpanWebClientCustomizerTest.kt]: Fixed with lint apply. https://github.com/yhware/sentry-java/commit/6eeee247c83bbae12250946aed6a0f292bd1282b#diff-f0f2e71f79903096ccffb59e7c6acf382f0bdace85f139f52ffd26db72f18feeL7

**Review by maciejwalkowiak — APPROVED** [review]
Thanks for fixes!

**Pull request merged**